hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataCompactResolver.java

```diff
@@ -0,0 +1,179 @@
+        }
+
+        return extractFields(userFields, isKey).entrySet().stream()
+                .map(entry -> {
+                    QueryPath path = entry.getKey();
+                    if (path.getPath() == null) {
+                        throw QueryException.error("Cannot use the '" + path + "' field with Compact serialization");
+                    }
+                    return entry.getValue();
+                });
+    }
+
+    Stream<MappingField> resolveFields(boolean isKey, Schema schema) {
+        if (schema.getFieldCount() == 0) {
+            // we didn't find any non-object fields, map the whole value
+            String name = isKey ? KEY : VALUE;
+            return Stream.of(new MappingField(name, QueryDataType.OBJECT, name));
+        }
+        return schema.getFieldNames().stream()
+                .map(name -> {
+                    QueryPath path = new QueryPath(name, isKey);
+                    QueryDataType type = resolveCompactType(schema.getField(name).getType());
+
+                    return new MappingField(name, type, path.toString());
+                });
+    }
+
+    @SuppressWarnings("checkstyle:ReturnCount")
+    private static QueryDataType resolveCompactType(FieldType type) throws IllegalStateException {
+        switch (type) {
+            case BOOLEAN:
+                return QueryDataType.BOOLEAN;
+            case BYTE:
+                return QueryDataType.TINYINT;
+            case SHORT:
+                return QueryDataType.SMALLINT;
+            case INT:
+                return QueryDataType.INT;
+            case LONG:
+                return QueryDataType.BIGINT;
+            case FLOAT:
+                return QueryDataType.REAL;
+            case DOUBLE:
+                return QueryDataType.DOUBLE;
+            case DECIMAL:
+                return QueryDataType.DECIMAL;
+            case CHAR:
+                return QueryDataType.VARCHAR_CHARACTER;
+            case UTF:
+                return QueryDataType.VARCHAR;
+            case TIME:
+                return QueryDataType.TIME;
+            case DATE:
+                return QueryDataType.DATE;
+            case TIMESTAMP:
+                return QueryDataType.TIMESTAMP;
+            case TIMESTAMP_WITH_TIMEZONE:
+                return QueryDataType.TIMESTAMP_WITH_TZ_OFFSET_DATE_TIME;
+            default:
+                return QueryDataType.OBJECT;
+        }
+    }
+
+    @Override
+    public KvMetadata resolveMetadata(
+            boolean isKey,
+            List<MappingField> resolvedFields,
+            Map<String, String> options,
+            InternalSerializationService serializationService
+    ) {
+        Map<QueryPath, MappingField> fieldsByPath = extractFields(resolvedFields, isKey);
+
+        String typeNameProperty = isKey ? OPTION_KEY_COMPACT_TYPE_NAME : OPTION_VALUE_COMPACT_TYPE_NAME;
+        String typeName = options.get(typeNameProperty);
+
+        if (typeName == null) {
```

Comment:
It should be validated in `resolveAndValidateFields`. Currently, `resolveMetadata` is called after mapping creation, during query execution for each mapping even if it's not used for given query.